projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/memberships.php

```diff
@@ -246,6 +287,130 @@ public function create_products( $request ) {
+		);
+
+		if ( ( defined( 'IS_WPCOM' ) && IS_WPCOM ) ) {
+			require_lib( 'memberships' );
+			$product = Memberships_Product::get_from_post( get_current_blog_id(), $product_id );
+			if ( is_wp_error( $product ) ) {
+				return array( 'error' => $product->get_error_message() );
+			}
+			if ( ! $product || ! $product instanceof Memberships_Product ) {
+				return array( 'error' => __( 'Product not found.', 'jetpack' ) );
+			}
+
+			$updated_product = $product->update( $payload );
+			if ( is_wp_error( $updated_product ) ) {
+				return array( 'error' => $updated_product->get_error_message() );
+			}
+			return array( 'product' => $updated_product->to_array() );
+		} else {
+			$payload = array(
+				'type'     => $request['type'],
+				'currency' => $request['currency'],
+			);
+
+			// If we pass directly is_editable as null, it would break API argument validation.
+			if ( null !== $is_editable ) {
+				$payload['is_editable'] = $is_editable;
+			}
+
+			$blog_id  = Jetpack_Options::get_option( 'id' );
+			$response = Client::wpcom_json_api_request_as_user(
+				"/sites/$blog_id/{$this->rest_base}/product/$product_id",
+				'v2',
+				array(
+					'method' => 'POST',
+				),
+				$payload
+			);
+			if ( is_wp_error( $response ) ) {
+				if ( $response->get_error_code() === 'missing_token' ) {
+					return new WP_Error( 'missing_token', __( 'Please connect your user account to WordPress.com', 'jetpack' ), 404 );
+				}
+				return new WP_Error( 'wpcom_connection_error', __( 'Could not connect to WordPress.com', 'jetpack' ), 404 );
+			}
+			$data = isset( $response['body'] ) ? json_decode( $response['body'], true ) : null;
+			// If endpoint returned error, we have to detect it.
+			if ( 200 !== $response['response']['code'] && $data['code'] && $data['message'] ) {
+				return new WP_Error( $data['code'], $data['message'], 401 );
+			}
+			return $data;
+		}
+
+		return $request;
+	}
+
+	/**
+	 * Delete an existing memberships product
+	 *
+	 * @param \WP_REST_Request $request The request passed from WP.
+	 *
+	 * @return array|WP_Error
+	 */
+	public function delete_product( \WP_REST_Request $request ) {
+		$product_id = $request->get_param( 'product_id' );
+		if ( ( defined( 'IS_WPCOM' ) && IS_WPCOM ) ) {
+			require_lib( 'memberships' );
+			$product = Memberships_Product::get_from_post( get_current_blog_id(), $product_id );
+			if ( is_wp_error( $product ) ) {
+				return array( 'error' => $product->get_error_message() );
+			}
+			if ( ! $product || ! $product instanceof Memberships_Product ) {
+				return array( 'error' => __( 'Product not found.', 'jetpack' ) );
+			}
+
+			$result = $product->delete();
+			if ( is_wp_error( $result ) ) {
+				return array( 'error' => $result->get_error_message() );
+			}
+			return array( 'deleted' => true );
+		} else {
+			$blog_id  = Jetpack_Options::get_option( 'id' );
+			$response = Client::wpcom_json_api_request_as_user(
+				"/sites/$blog_id/{$this->rest_base}/product/$product_id/delete",
```

Comment:
Here you should use "/sites/$blog_id/{$this->rest_base}/product/$product_id" and DELETE method, as defined https://github.com/Automattic/wp-calypso/pull/80205/files#diff-5e4abdd9e35ec1230c13cf42dcc2723cc52fd5025ec8e0be5370fcd736eed451R135

Also, it seems like you don't defined the DELETE endpoint here. I will try now, but not sure how this diff could work as is with the Calypso fix 🤔

Reply:
Thanks, I'll refactor this. I think v1 used the `/delete` syntax, but you're right that v2 uses the `DELETE` method instead.

I think the tests were working because this block of code isn't actually hit in this flow - we can only delete products from calypso at the moment, which means `IS_WPCOM` will be true. This code was included so that we can proxy to WPCOM in the future if we add delete functionality to the gutenberg blocks and panels. As such, I don't have a manual test for this block yet.

Reply:
This `delete_product` endpoint is registered around line 155 in this file:

```php
			array(
					'methods'             => WP_REST_Server::DELETABLE,
					'callback'            => array( $this, 'delete_product' ),
					'permission_callback' => array( $this, 'get_status_permission_check' ),
				),
```

Reply:
> This `delete_product` endpoint is registered around line 155 in this file:

Sorry. For some reason I totally missed it...